### PR TITLE
DROOLS-2298: Inline controls: Function editor

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/DROOLS-2298

Video [here](https://youtu.be/8x0H-0zIz-E). UI for "Parameter popup" to be finalised by UX.

### CODE REVIEW

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/FunctionColumnParametersHeaderMetaData.java [hunk 1]
```diff
@@ -56,13 +67,13 @@ public String getTitle() {
...
+    String getExpressionLanguageTitle() {
+        return KindUtilities.getKind(functionSupplier.get()).code();
     }
 
-    public String getFormalParametersTitle() {
-        final List<InformationItem> formalParameters = formalParametersSupplier.get();
-        final StringBuffer sb = new StringBuffer();
+    String getFormalParametersTitle() {
+        final List<InformationItem> formalParameters = functionSupplier.get().getFormalParameter();
+        final StringBuilder sb = new StringBuilder();
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/FunctionColumnParametersHeaderMetaData.java]: @manstis Would you mind to apply this change also in the `getTitle()` method, or do we need the synchronized `StringBuffer` there?

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/FunctionColumnParametersHeaderMetaData.java [hunk 2]
```diff
@@ -75,4 +86,31 @@ public String getFormalParametersTitle() {
...
+
+        editor.bind(gridWidget,
+                    uiRowIndex,
+                    uiColumnIndex);
+        final double[] dxy = {0.0, 0.0};
+        final Optional<Point2D> rx = context.getRelativeLocation();
+        rx.ifPresent(r -> {
+            final double headerRowHeight = context.getCellHeight();
+            dxy[1] = r.getY() - headerRowHeight * uiRowIndex;
+            dxy[0] = r.getX();
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/FunctionColumnParametersHeaderMetaData.java]: Wouldn't be more readable the code below?
```java
dxy[0] = ...
dxy[1] = ...
```

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties [hunk 3]
```diff
@@ -36,8 +36,11 @@ RelationEditor.DeleteColumn=Delete column
 InvocationEditor.InsertParameterAbove=Insert Parameter above
 InvocationEditor.InsertParameterBelow=Insert Parameter below
 InvocationEditor.DeleteParameter=Delete Parameter
+FunctionEditor.FEEL=FEEL
+FunctionEditor.JAVA=JAVA
+FunctionEditor.PMML=PMML
+ParametersEditorViewImpl.addParameter=Add parameter
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties]: Same question as last time, shouldn't we use `Insert parameter` due to other constants in inline context menus?

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties]: This isn't on a context menu but a button on the "Parameters Popup". The User is also not _inserting_ a Parameter (above or below) but _adding_ (or even _appending_) a new one to the list. I can change it if you like but the whole "Parameters Popup" is yet to be designed by UX - this one is something I put together to fill the need.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/FunctionColumnParametersHeaderMetaData.java (hunk 2 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/FunctionColumnParametersHeaderMetaData.java]: Done (in new commit to follow!)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/FunctionColumnParametersHeaderMetaData.java (hunk 1 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/FunctionColumnParametersHeaderMetaData.java]: Updated all use of ```StringBuffer``` to ```StringBuilder``` in the code base (in new commit to follow!)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/FunctionGrid.java [hunk 4]
```diff
@@ -276,34 +323,15 @@ void doSetKind(final FunctionDefinition.Kind kind,
...
-                                                                  () -> synchroniseViewWhenExpressionEditorChanged(editor)));
-        });
-    }
-
-    @Override
-    protected void synchroniseViewWhenExpressionEditorChanged(final Optional<BaseExpressionGrid> oEditor) {
-        super.synchroniseViewWhenExpressionEditorChanged(oEditor);
-        getEditorControls();
+    void clearExpressionType() {
+        final GridCellTuple gc = new GridCellTuple(0,
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/FunctionGrid.java]: Shouldn't start this as `expression.ifPresent(function...`?

**Comment by manstis** [comment]
@jomarko OK, all issues (sort of) resolved!

There is a bug in the backend marshalling I need to raise a JIRA for @tarilabs to look into regarding marshalling and unmarshalling of ```QName```. I added ```equals(..)``` and ```hashCode()``` on ```QName``` as it is being used as the key in a ```Map```.

**Comment by manstis** [comment]
@jomarko It is this ```QName``` marshalling/unmarshalling bug that prevents ```FunctionDefinition``` from being re-opened.

**Comment by jomarko** [comment]
Ok sorry, I got it now, so I won't be able to reopen within this PR right?

**Comment by manstis** [comment]
@jomarko Correct; until https://issues.jboss.org/browse/DROOLS-2372 is resolved (or any other misunderstanding about how QNames are meant to be working between DMN XML and the WB) creating a Function, saving, closing and re-opening is broken.

**Comment by manstis** [comment]
@jomarko Last issue you report with "empty" values is not related to ```QName``` :-)

When a cell value is an "empty string" the cell is being deleted rather than then value cleared.

Could you please open a new JIRA for that?

**Comment by manstis** [comment]
@jomarko ... and it wasn't "the exception as previously"; it was "Cannot read property 'getValue_2_g$' of null"

**Comment by manstis** [comment]
Thanks @jomarko I'll update my other PRs.

**Pull request merged**